tools/mascot/src/main/java/dk/alexandra/fresco/tools/mascot/triple/TripleGeneration.java

```diff
@@ -0,0 +1,307 @@
+import dk.alexandra.fresco.tools.mascot.mult.MultiplyLeft;
+import dk.alexandra.fresco.tools.mascot.mult.MultiplyRight;
+import dk.alexandra.fresco.tools.mascot.utils.FieldElementPrg;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+/**
+ * Actively-secure protocol for computing authenticated, secret-shared multiplication triples based
+ * on the MASCOT protocol (https://eprint.iacr.org/2016/505.pdf).<br>
+ * In particular, produces random, authenticated, secret-shared triples of the form a, b, c such
+ * that a * b = c.
+ */
+public class TripleGeneration extends BaseProtocol {
+
+  private final ElementGeneration elementGeneration;
+  private final Map<Integer, MultiplyRight> rightMultipliers;
+  private final Map<Integer, MultiplyLeft> leftMultipliers;
+  private final FieldElementPrg jointSampler;
+
+  /**
+   * Creates new triple generation protocol.
+   */
+  public TripleGeneration(MascotResourcePool resourcePool, Network network,
+      ElementGeneration elementGeneration, FieldElementPrg jointSampler) {
+    super(resourcePool, network);
+    this.leftMultipliers = new HashMap<>();
+    this.rightMultipliers = new HashMap<>();
+    for (Integer partyId : getPartyIds()) {
+      if (!partyId.equals(getMyId())) {
+        rightMultipliers.put(partyId, new MultiplyRight(resourcePool, network, partyId));
+        leftMultipliers.put(partyId, new MultiplyLeft(resourcePool, network, partyId));
+      }
+    }
+    this.elementGeneration = elementGeneration;
+    this.jointSampler = jointSampler;
+  }
+
+  TripleGeneration(MascotResourcePool resourcePool, Network network, FieldElementPrg jointSampler,
+      FieldElement macKeyShare) {
+    this(resourcePool, network,
+        new ElementGeneration(resourcePool, network, macKeyShare, jointSampler), jointSampler);
+  }
+
+  List<UnauthTriple> toUnauthTriple(List<FieldElement> left, List<FieldElement> right,
+      List<FieldElement> prods) {
+    Stream<UnauthTriple> stream = IntStream.range(0, right.size()).mapToObj(idx -> {
+      int groupStart = idx * getNumCandidatesPerTriple();
+      int groupEnd = (idx + 1) * getNumCandidatesPerTriple();
+      return new UnauthTriple(left.subList(groupStart, groupEnd), right.get(idx),
+          prods.subList(groupStart, groupEnd));
+    });
+    return stream.collect(Collectors.toList());
+  }
+
+  List<FieldElement> multiply(List<FieldElement> leftFactorGroups,
+      List<FieldElement> rightFactors) {
+    // "stretch" right factors, so we have one right factor for each left factor
+    List<FieldElement> stretched =
+        getFieldElementUtils().stretch(rightFactors, getNumCandidatesPerTriple());
+
+    // for each value we will have two sub-factors for each other party
+    List<List<FieldElement>> subFactors = new ArrayList<>();
+
+    // TODO parallelize
```

Comment:
Is this still a TODO or irellevant now?

Reply:
This is a performance improvement for the > 2 party case. I will remove the TODO and make an improve performance issue instead where this can go as a subtask.